packages/widget/src/controllers/transfers/evm/build.ts

```diff
@@ -1,89 +1,113 @@
-  this: FungibleTokenTransferController
-): Promise<void> {
-  //we already check that but this prevents those typescript errors
-  const provider = this.walletContext.value?.evmWallet?.provider;
-  const providerChaiId = this.walletContext.value?.evmWallet?.providerChainId;
-  const address = this.walletContext.value?.evmWallet?.address;
-  if (
-    !this.sourceNetwork ||
-    !this.destinationNetwork ||
-    !this.resourceAmount ||
-    !this.selectedResource ||
-    !this.destinationAddress ||
-    !provider ||
-    !address ||
-    providerChaiId !== this.sourceNetwork.chainId
-  ) {
-    this.estimatedGas = undefined;
-    this.resetFee();
-    return;
-  }
-
+export async function buildEvmFungibleTransactions({
+  address,
+  chainId,
+  destinationAddress,
+  resourceId,
+  resourceAmount,
+  provider,
+  providerChainId,
+  env,
+  pendingEvmApprovalTransactions,
+  pendingTransferTransaction,
+  fee
+}: {
+  address: string;
+  chainId: number;
+  destinationAddress: string;
+  resourceId: string;
+  resourceAmount: BigNumber;
+  provider: Eip1193Provider;
+  providerChainId: number;
+  env: Environment;
+  pendingEvmApprovalTransactions: UnsignedTransaction[];
+  pendingTransferTransaction: UnsignedTransaction;
+  sourceNetwork: Domain | null;
+  fee: EvmFee;
+}): Promise<{
+  pendingEvmApprovalTransactions: UnsignedTransaction[];
+  pendingTransferTransaction: UnsignedTransaction;
+  fee: EvmFee;
+  resourceAmount: BigNumber;
+}> {
   const evmTransfer = new EVMAssetTransfer();
-  await evmTransfer.init(new Web3Provider(provider, providerChaiId), this.env);
+  await evmTransfer.init(new Web3Provider(provider, providerChainId), env);
 
   // Hack to make fungible transfer behave like it does on substrate side
   // where fee is deducted from user inputted amount rather than added on top
   const originalTransfer = await evmTransfer.createFungibleTransfer(
     address,
-    this.destinationNetwork.chainId,
-    this.destinationAddress,
-    this.selectedResource.resourceId,
-    this.resourceAmount.toString()
+    chainId,
+    destinationAddress,
+    resourceId,
+    resourceAmount.toString()
   );
   const originalFee = await evmTransfer.getFee(originalTransfer);
+  // NOTE: for percentage fee, if both are equal, it means we can calculate the amount with fee avoiding second subtraction
+  const calculateAmountWithFee = originalFee.type === FeeHandlerType.PERCENTAGE;
+
   //in case of percentage fee handler, we are calculating what amount + fee will result int user inputed amount
   //in case of fixed(basic) fee handler, fee is taken from native token
-  if (originalFee.type === FeeHandlerType.PERCENTAGE) {
+  if (calculateAmountWithFee) {
     const { lowerBound, upperBound, percentage } = originalFee as PercentageFee;
-    const userInputAmount = this.resourceAmount;
+    const userInputAmount = resourceAmount;
     //calculate amount without fee (percentage)
     const feelessAmount = userInputAmount
```

Comment:
Sounds weird :) what about the more boring version: "amountWithoutFee"?

Reply:
i also feel there should a function that calculates this stuff that way we can also add tests for that

Reply:
@Lykhoyda done with your suggestion. @saadjhk could you create an issue for this?